For docker stack ls make an error if Swarm and Kubernetes hosts do not match

mat007 opened the pull request:
**- What I did**

Issued an error for docker stack ls if Swarm and Kubernetes hosts do not match when ```--orchestrator=all```

**- How I did it**

By comparing the host names:
* if the host names are the same then we're good
* if the docker daemon is reached using a scheme different than tcp (e.g. fd, npipe or unix) and Kubernetes host name/address is a loopback, we're good too
* otherwise report an error

**- How to verify it**

Having both Docker for Desktop with Kubernetes and UCP installed and configured:
```
$ kubectl config use-context docker-for-desktop
Switched to context "docker-for-desktop".

$ docker stack ls --orchestrator=all
NAME                SERVICES            ORCHESTRATOR        NAMESPACE
dtc                 5                   Kubernetes          tata
dtc                 5                   Kubernetes          tutu
wall-e              2                   Swarm

$ kubectl config use-context ucp_34.209.72.126:6443_admin
Switched to context "ucp_34.209.72.126:6443_admin".

$ docker stack ls --orchestrator=all
Swarm and Kubernetes hosts do not match, check DOCKER_HOST and Kubernetes context
```

**- Description for the changelog**

* Issue an error for --orchestrator=all when working on mismatched Swarm and Kubernetes hosts

**- A picture of a cute animal (not mandatory but encouraged)**

![image](https://user-images.githubusercontent.com/2386884/39583414-5c99847a-4ef0-11e8-9057-3e0ada05cf8f.png)

mat007 commented:
🍂 comes after #1031, only the last commit is new.

thaJeztah commented:
Renamed to "WIP", because this depends on https://github.com/docker/cli/pull/1031 to be merged first

thaJeztah commented:
Ah, yes, I was looking at this one a couple of times, trying to grasp the "local socket" situation; let me have a look again

silvin-lubecki commented on cli/command/stack/kubernetes/cli.go:
```diff
@@ -75,3 +84,30 @@ func WrapCli(dockerCli command.Cli, opts Options) (*KubeCli, error) {
+			}
+		}
+	}
+	return fmt.Errorf("Swarm and Kubernetes hosts do not match, check DOCKER_HOST and Kubernetes context")
```
Maybe it should land to `docker info` instead?

thaJeztah replied:
Yes, docker info definitely, and likely the error/warning should include the information as well;

    Swarm and Kubernetes hosts do not match, check DOCKER_HOST and Kubernetes context

(i.e, it should mention what it found)

dhiltgen replied:
I agree we should try to get more details exposed (most likely in `info`) but there may be some other wrinkles we need to work out to get that fully wired up.  A follow up PR perhaps?

If we could improve the error message to give a little more contextual information that might help users get themselves unstuck.  Perhaps something like this?
```go
return fmt.Errorf("Swarm and Kubernetes hosts do not match - DOCKER_HOST=%s and kubernetes context=%s - update DOCKER_HOST or use 'kubectl config use-context' to match", daemonEndpoint.Hostname(), kubeEndpoint.Hostname())
```

dhiltgen replied:
I think folks were concerned about subtle failure modes with miswired configurations leading to worse confusion.  My $0.02 is a hard fail is OK as long as there's enough detail in the error message so the user can figure it out and correct the problem.

thaJeztah replied:
> so the user can figure it out and correct the problem.

That's exactly the problem: the user's configuration is correct, so there _is no workaround_ for the user.

It's a "best effort" detection to notify the user in situations where (possibly) swarm and kubernetes don't match, but there's many scenarios where we will be wrong.

mat007 commented:
I updated the message and changed it into a warning log, @thaJeztah @dhiltgen just tell me if we rather want an error instead and I'll change it back.

What sort of output do we want in `docker info`?
There's already a `Swarm` section but nothing for `Kubernetes` so far.
Or do we rather have a separate `Orchestrators` section to group them together?

thaJeztah commented on cli/command/stack/kubernetes/cli.go:
```diff
@@ -69,9 +73,43 @@ func WrapCli(dockerCli command.Cli, opts Options) (*KubeCli, error) {
+			}
+		}
+	}
+	fmt.Fprintf(c.Out(), "Warning: Swarm and Kubernetes hosts do not match - DOCKER_HOST=%s and kubernetes context=%s -"+
```
Should probably be printed on `stderr` instead of `stdout`?

thaJeztah commented on cli/command/stack/kubernetes/cli.go:
```diff
@@ -69,9 +73,43 @@ func WrapCli(dockerCli command.Cli, opts Options) (*KubeCli, error) {
+		}
+	}
+	fmt.Fprintf(c.Out(), "Warning: Swarm and Kubernetes hosts do not match - DOCKER_HOST=%s and kubernetes context=%s -"+
+		" update DOCKER_HOST or use 'kubectl config use-context' to match\n", daemonEndpoint.Hostname(), kubeEndpoint.Hostname())
```
let me think a bit about mentioning `DOCKER_HOST`, because this could also be because of `docker -H=<some host>`

mat007 replied:
What about
```
"Warning: Swarm and Kubernetes hosts do not match - docker host=%s and kubernetes context=%s - update DOCKER_HOST (or pass -H) or use 'kubectl config use-context' to match```
?

thaJeztah replied:
Yes that looks better. Perhaps put the "docker host / kubernetes host" in brackets, and align the output for readability (as it's printed when working with docker interactively, we should make it user-friendly) e.g.;


```
WARNING: Swarm and Kubernetes hosts do not match (docker host=%s, kubernetes host=%s)
         Update $DOCKER_HOST (or pass -H), or use 'kubectl config use-context' to match.
```

Should we use `context=` or `host=` for k8s? "Context" may be confused with the name of the context?

mat007 replied:
`host` should be fine, I updated the message.
